dizoo/board_games/gomoku/envs/gomoku_env.py

```diff
@@ -0,0 +1,253 @@
+        else:
+            print("Error: input illegal action, we randomly choice a action from self.legal_actions!")
+            action = np.random.choice(self.legal_actions)
+            row, col = self.action_to_coord(action)
+            self.board[row, col] = self.current_player
+            self._current_player = self.current_opponent_player
+            # sys.exit(-1)
+
+        done, winner = self.have_winner()
+
+        reward = int((winner == curr_player))
+        info = {'next player to play': next_player}
+
+        if done:
+            self._final_eval_reward = reward
+            info['final_eval_reward'] = self._final_eval_reward
+            print('gomoku one episode done: ', info)
+
+        action_mask = np.zeros(self.num_actions, 'int8')
+        action_mask[self.legal_actions] = 1
+        obs = {'observation': self.current_state(), 'action_mask': action_mask}
+        return BaseEnvTimestep(obs, reward, done, info)
+
+    def current_state(self):
+        board_curr_player = np.where(self.board == self.current_player, 1, 0)
+        board_opponent_player = np.where(self.board == self.current_opponent_player, 1, 0)
+        board_to_play = np.full((self.board_size, self.board_size), self.current_player)
+        return np.array([board_curr_player, board_opponent_player, board_to_play], dtype=np.float32)
+
+    def coord_to_action(self, i, j):
+        ''' convert coordinate i, j to action a in [0, board_size**2)
+        '''
+        a = i * self.board_size + j  # action index
+        return a
+
+    def action_to_coord(self, a):
+        coord = (a // self.board_size, a % self.board_size)
+        return coord
+
+    def have_winner(self):
+        has_legal_actions = False
+        directions = ((1, -1), (1, 0), (1, 1), (0, 1))
+        for i in range(self.board_size):
+            for j in range(self.board_size):
+                # if no stone is on the position, don't need to consider this position
+                if self.board[i][j] == 0:
+                    has_legal_actions = True
+                    continue
+                # value-value at a coord, i-row, j-col
+                player = self.board[i][j]
+                # check if there exist 5 in a line
+                for d in directions:
+                    x, y = i, j
+                    count = 0
+                    for _ in range(5):
+                        if (x not in range(self.board_size)) or (
+                                y not in range(self.board_size)
+                        ):
+                            break
+                        if self.board[x][y] != player:
+                            break
+                        x += d[0]
+                        y += d[1]
+                        count += 1
+                        # if 5 in a line, store positions of all stones, return value
+                        if count == 5:
+                            return True, player
+        return not has_legal_actions, -1
```

Comment:
maybe adding a comment that if no position to action, return done=True is more explicit?